pkg/stanza/fileconsumer/internal/reader/reader.go

```diff
@@ -209,15 +212,47 @@ func (r *Reader) readContents(ctx context.Context) {
 			r.FileAttributes[attrs.LogFileRecordNumber] = r.RecordNum
 		}
 
-		err = r.emitFunc(ctx, emit.NewToken(token, r.FileAttributes))
-		if err != nil {
-			r.set.Logger.Error("failed to process token", zap.Error(err))
+		tokens = append(tokens, emit.NewToken(copyBody(token), copyAttributes(r.FileAttributes)))
```

Comment:
What specifically introduces a need to make copies of the body and attributes?

It seems like a lot of overhead - are we sure it's necessary?

Reply:
Without copying, for each token in the `tokens` collection, its Body would point to the same place in memory, specifically the Decoder's [decodeBuffer](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.115.0/pkg/stanza/decode/decode.go#L21).

This isn't a problem when the Reader emits each token before creating a new token, as the token's Body is copied in the `emit` function (specifically in the [toBody](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.115.0/pkg/stanza/operator/input/file/config.go#L48) function - the conversion from `[]byte` to `string` [copies](https://github.com/golang/go/issues/25484) the underlying array).

Similar with the attributes, the `reader::FileAttributes` map would be reused by all tokens if not copied.

Reply:
Oh, I think you're right. I think I assumed that this can also happen when the context is cancelled (e.g. when collector receives SIGINT). But when the context is canceled, the method `return`s instead of `break`ing out of the loop, which I suppose is an OK thing to do. I'll take another look at it.

Reply:
If I'm understanding correctly then, we're just making the copies in a different place, so this does not introduce new work into the hot path?

Reply:
I'm afraid it does introduce new work.

How I believe this worked before this change:

- Every time a new token is read from the file, it's placed in the same area in memory - the Decoder's [decodeBuffer](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.115.0/pkg/stanza/decode/decode.go#L21).
- A pointer to that place in memory is passed to the `emit` function
- The `emit` function reads the bytes and creates a new `Entry`, copying the bytes into a new area in memory.

How this works after this change:

- Every time a new token is read from the file, it's placed in the same area in memory - the Decoder's [decodeBuffer](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.115.0/pkg/stanza/decode/decode.go#L21).
- In the Reader's `readContent` function, a batch of tokens is accumulated before calling the `emit` function, so the bytes from `decodeBuffer` need to be copied into a new area in memory for every new token
- the accumulated slice of tokens is passed to the `emit` function, which creates a new Entry for every Token, copying the bytes for the second time.